### PR TITLE
Improve accessibility of the Warning component in the block editor

### LINKED ISSUE #62737: Improve the Block editor warning accessibility and design
### Description

The block editor [Warning](https://github.com/WordPress/gutenberg/blob/d803f99b6a6613c16cffb9f639398ca8f3b17bf5/packages/block-editor/src/components/warning/index.js) component can be improved in terms of accessibility and design.

1
When the Warning renders in the editor, it's only a _visual_ warning. There's no `speak` message or ARIA `role=alert` that would make the warning message be automatically announced by screen readers.

2
The design could be improved. Screenshot:

Screenshot 2024-06-21 at 12 14 02

### Step-by-step reproduction instructions

I tested one of the implementation of the Warning, that is the [MultipleUsageWarning](https://github.com/WordPress/gutenberg/blob/d803f99b6a6613c16cffb9f639398ca8f3b17bf5/packages/block-editor/src/components/block-edit/multiple-usage-warning.js) component:

- Create a post and add a More block.
- Select the block and Copy either via the keyboard Copy shortcut or via the equivalent command in the block toolbar Options menu.
- Click the bottom part of the post to get the default block.
- Paste the copied More block.
- Observe the MultipleUsageWarning appears.
- Observe there's no screen reader announcement.
- Observe the Warning design could be improved.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by sarthaknagoshe2002**
Accessibility Fix: #62737

## What?
This PR improves the Warning component in the block editor by enhancing its accessibility.

## Why?
Currently, the Warning component only provides a visual cue without notifying screen readers, which limits accessibility. These changes ensure that the component is accessible to all users.

## How?

- Added ARIA role="alert" and aria-live="assertive" to the Warning component for screen reader announcements.
- Added aria-describedby="block-editor-warning-message" to the Warning actions.
- Added tabIndex="0", so that the Warning can be accessed using tabs. This helps to navigate inside the warning.

## Testing Instructions

1. Add a More block to a post and copy it using the keyboard shortcut or the Options menu.
2. Paste the copied block elsewhere in the post.
3. Observe the Warning message appears with improved design and triggers a screen reader announcement.

### CODE REVIEW

**Comment by afercia** [comment]
@sarthaknagoshe2002 thanks for your PR.

Good catch for adding the `tabindex="0"` which makes possible to navigate into the Warning with the keyboard.

Unfortunately, testing with Safari and VoiceOver the Warning isn't announced. I guess because the editor sets focus on the newly inserted block. See in the screenshot below: focus is on the doubled 'More' block:

Screenshot 2024-12-17 at 09 07 55

I think a potential way to fix this would be setting focus on the Warning when it renders by using `useEffect` and `useRef`. 

I'll leave a few comments in the review as well. 
Worth noting there's a merge conflict because it appears this branch doesn't incorporate recent changes to the Warning component introduced in https://github.com/WordPress/gutenberg/pull/67675

File: packages/block-editor/src/components/warning/index.js [hunk 1]
```diff
@@ -14,14 +14,22 @@ import { moreVertical } from '@wordpress/icons';
 function Warning( { className, actions, children, secondaryActions } ) {
 	return (
 		<div style={ { display: 'contents', all: 'initial' } }>
-			<div className={ clsx( className, 'block-editor-warning' ) }>
+			<div
+				className={ clsx( className, 'block-editor-warning' ) }
+				role="alert"
+				aria-live="assertive"
```
Comment (afercia) [inline packages/block-editor/src/components/warning/index.js]: role alert inherently works as an assertive live region so that `aria-live="assertive"` isn't necessary. `aria-live` doesn;t work well for elements that appear in the DOM anyways.

File: packages/block-editor/src/components/warning/index.js [hunk 2]
```diff
@@ -14,14 +14,22 @@ import { moreVertical } from '@wordpress/icons';
...
 				<div className="block-editor-warning__contents">
 					<p className="block-editor-warning__message">
 						{ children }
 					</p>
 
 					{ ( Children.count( actions ) > 0 || secondaryActions ) && (
-						<div className="block-editor-warning__actions">
+						<div
+							className="block-editor-warning__actions"
+							aria-describedby="block-editor-warning-message"
```
Comment (afercia) [inline packages/block-editor/src/components/warning/index.js]: Not sure this `aria-describedby` is necessary. It doesn't point to any ID anyways, please remove it.

**Review by afercia — CHANGES_REQUESTED** [review]
Left some comments

**Comment by sarthaknagoshe2002** [comment]
> I think a potential way to fix this would be setting focus on the Warning when it renders by using `useEffect` and `useRef`.

It seems to work fine in Chrome on macOS using the default screen reader. I also tested it on Safari, and it’s a bit buggy—I noticed it works intermittently. I’m currently investigating the issue and exploring possible solutions.

**Comment by sarthaknagoshe2002** [comment]
@afercia I have now fixed things to make it work in safari.
Also, I have implemented the changes suggested by you.

| Chrome | Safari |
|-|-|
| Screenshot 2024-12-17 at 11 04 03 PM | Screenshot 2024-12-17 at 11 42 21 PM |


Let me know this this requires any more changes. If not then I'll update the tests too. 😇

**Comment by sarthaknagoshe2002** [comment]
Also, the git hook seems to have modified the linting of some tsconfig.json file.
let me know if we should keep those changes too.

File: packages/block-editor/src/components/warning/index.js [hunk 3]
```diff
@@ -9,13 +9,28 @@ import clsx from 'clsx';
...
+import { useEffect, useRef } from '@wordpress/element';
 
 function Warning( { className, actions, children, secondaryActions } ) {
+	const alertRef = useRef( null );
+
+	useEffect( () => {
+		if ( null !== alertRef.current ) {
+			alertRef.current.focus();
+		}
+	}, [] );
```
Comment (afercia) [inline packages/block-editor/src/components/warning/index.js:19]: Question: can this be simplified? I'm not sure there's the need to set the initial ref to `null` and then checking for it. To my understanding the optional chaining operator `?.` would make this much more concise. See an example here:

https://github.com/WordPress/gutenberg/blob/e4649fe6896123231de35da0875e93ac2920c23a/packages/block-library/src/video/tracks-editor.js#L182-L186

File: packages/upload-media/tsconfig.json [hunk 4]
```diff
@@ -8,13 +8,13 @@
 	},
 	"include": [ "src/**/*" ],
 	"references": [
-	  { "path": "../api-fetch" },
```
Comment (afercia) [inline packages/upload-media/tsconfig.json]: changes to this file should be reverted, probably a result of merge conflict or something

File: packages/block-editor/src/components/warning/index.js [hunk 5]
```diff
@@ -9,13 +9,28 @@ import clsx from 'clsx';
...
-			<div className={ clsx( className, 'block-editor-warning' ) }>
+			<div
+				className={ clsx( className, 'block-editor-warning' ) }
+				tabIndex="0"
+			>
 				<div className="block-editor-warning__contents">
-					<p className="block-editor-warning__message">
+					<p
+						className="block-editor-warning__message"
+						ref={ alertRef }
```
Comment (afercia) [inline packages/block-editor/src/components/warning/index.js]: I think the reg should be set on the container with tabIndex="0", otherwise moving focus will fail?

**Comment by afercia** [comment]
> let me know if we should keep those changes too.

@sarthaknagoshe2002 yes I think those changes should be reverted.
Thanks for working on this, I left some new comments.

As a last step, when everything is done, I guess the snapshot of the failing test should be updated.

```
npm run test:unit -- --updateSnapshot --testPathPattern packages/block-editor/src/components/warning/test/index.js
```

**Comment by afercia** [comment]
It's worth noting that _all_ Warnings will receive focus when they render, for example the 'invalid content' Warning and other ones. There may be edge cases but I think it's OK as default behavior.

Screenshot 2024-12-18 at 09 19 18

File: packages/upload-media/tsconfig.json (hunk 4 above)
Comment (sarthaknagoshe2002) [reply packages/upload-media/tsconfig.json]: The pre-commit Git hook is enforcing linting even after reverting changes.  Are there alternative solutions?

Comment (afercia) [reply packages/upload-media/tsconfig.json]: I submitted a quick PR https://github.com/WordPress/gutenberg/pull/68083 to fix it separately.
Once that is merged, please rebase this PR.

**Comment by sarthaknagoshe2002** [comment]
@afercia I believe this is now ready to be merged.

**Review by afercia — APPROVED** [review]
Looks good to me, thanks @sarthaknagoshe2002 !

**Comment by afercia** [comment]
@WordPress/gutenberg-core as mentioned earlier, with this change _all_ Warnings will receive focus when they render. There may be edge cases but I think it's OK as default behavior and I'm going to merge this PR. Just wanted to be sure everyone is aware of this change in case there are edge cases to address in follow-ups.

**Comment by t-hamano** [comment]
I think having this component always receive focus when rendered can cause other accessibility issues.

For example, create a pattern that already has broken blocks. When that pattern is rendered via a site editor or similar, the `Warning` component in the pattern will be focused, causing the current navigation focus to be lost. And once focus is moved into an iframe, it cannot be moved out of it:

https://github.com/user-attachments/assets/79b081af-0bf8-4ee5-86bc-1b9ac428032d

Another concern is that when I open a post with multiple broken blocks, the last Warning component gets focus immediately. I would like to confirm if this is an issue or not.

**Comment by afercia** [comment]
@t-hamano thanks for your feedback. To me, moving focus is the most sensible default, especially when the Warning contains action buttons. Keyboard users will benefit from being able to immediately access the actions.

> causing the current navigation focus to be lost

Yes I see your point. Should Warnings be rendered in the Patterns preview card in the first place? I didn't know of this behavior but to me it looks pretty weird. That is a preview, it should not contain interactive elements and they're so small anyways that I'd argue they're not usable at all. Screenshot:

Screenshot 2024-12-19 at 10 31 50

> when I open a post with multiple broken blocks, the last Warning component gets focus immediately.

Ideally, there should be only one Warning at a time. If the editor shows _multiple_ warnings at the same time, they should also be _all announced_ to screen reader users to provide an equivalent information. However, implementing a way to announce _all_ warnings at the same time could be potentially very noisy and confusing. I'm really not sure how to address this potential edge case.

**Comment by afercia** [comment]
I'll create a separate issue for the Warnings in the preview cards.

**Comment by Mamaduka** [comment]
I agree with @t-hamano that the `Warning` component shouldn't autofocus because of rendering issues. We should revert it.

## Why?

* The "focus because of render" is a tricky pattern in React; it's hard to get right, and because of that, it should be avoided.
* Multiple blocks can render a `Warning` component since they block a specific. The last one to render will steal the focus.
* These are caused by invalid or broken content when a post is rendered, not by user action. So they can be handled later.

TL;DR, let's avoid making similar changes to components that can be rendered multiple times and are widely used by block editors.

**Comment by getdave** [comment]
I would also be in favour of reverting this. Autofocus on render is very likely to cause a number of knock on problems.

The other changes to make the warnings perceivable and more prominent to assistive tech seem like good improvements however.

**Comment by afercia** [comment]
I'm not opposed to a revertl. However:

> The other changes to make the warnings perceivable and more prominent to assistive tech seem like good improvements however.

- The other improvements don't work without the focus being set on the Warning.
- The Warning may provide (and in most of the cases it does) action buttons. It's pointless and non-accessible rendering action buttons if keyboard users have to tab dozens and dozens of times to navigate to them.
- The Warning is a sort of 'alert' message. To make it usable for everyone it should be semantically an alert and should be announced by assistive tech and be easily navigable to with keyboard. Ideally, as with all alerts, there should be only one Warning rendered at a time.
- For accessibility, it's not uncommon to move focus to alerts and warning messages. It's a pretty common pattern. rather then removing this behavior we should find a way to make it solid and reliable.

Overall, the _usage_ of the Warning is highly problematic and all these points should be addressed. If anyone can think at better ways to address these points, that would be very welcome.

**Comment by Mamaduka** [comment]
You can say that these warnings are mostly part of the content/canvas. Let's say you opened a documentation page with multiple warnings for experimental/unstable APIs. What are the expectations here? Do each of these warning messages try and move focus on itself, or are they just read as you get closer to actual experimental/unstable sections?

> rather then removing this behavior we should find a way to make it solid and reliable.

Ideally, we would have caught this broken behavior before merging this PR, but we are humans and sometimes miss things.

I'll include a full revert in my PR (#68133). Let's reopen the issue and continue working on a solution.

**Comment by afercia** [comment]
Was thinking at a potential different approach, I will comment on the issue and reopen it.

**Comment by Mamaduka** [comment]
Thanks, @afercia!

Meanwhile, my revert PR is ready to be merged - https://github.com/WordPress/gutenberg/pull/68133.

**Comment by sarthaknagoshe2002** [comment]
Since the focus was causing issues, we could have simply removed that part, as the warning was being announced by the screen reader in other browsers (except Safari) without the focus. This way, the rest of the functionality remains untouched.
